grammar.js

```diff
@@ -386,10 +387,6 @@ module.exports = grammar({
 	},
 });
 
-function preprocessor(command) {
-	return alias(new RegExp('#[ \t]*' + command), '#' + command);
-}
```

Comment:
This function was causing an error when I added support for `#include` in nodes. Interestingly enough, when I remove it, everything works as expected and all tests pass.  What's the purpose of this anyway?

Reply:
It is valid to have spaces between the `#` and the following command, e.g.
```c
#ifdef FOO
#    include "foo.h"
#endif
```
This function exists to allow parsing `#    include` and rename it to `#include` on the parse tree.

Reply:
I found that updating the regex to include a space after the preprocessor name worked.

```js
alias(new RegExp('#[ \t]*' + command + ' '), '#' + command);
```

Reply:
I think that would disallow other whitespace (e.g. tab) that should be valid there?

Reply:
Updated the regex to allow spaces and tabs between the directive and the argument.

Reply:
I think tree-sitter takes care of handling whitespace between tokens, I'm able to parse preproc statements with `\t` after the command part in the current devicetree grammar. Is there a specific case that it handles?

Reply:
The way the grammar is currently written, it doesn't properly parse preproc directives inside nodes or other non-toplevel statements. I'm not entirely sure why TBH, but it required these changes to make it work.

Reply:
Ok, I see that the new preprocessor test has a problem with it.

I guess something like `#foo` could look like a property name. If it were actually a conflict tree-sitter would fail to build the grammar but I wonder if appending the `[ \t]` to the pattern is cluing it in in a different way.

Moving the preprocessor stuff up in the choices list doesn't help, would the precedence functions be applicable here?

Reply:
I tried the precedence functions, but honestly I'm not sure if I was doing it right or wrong.

Reply:
Preprocessor commands should probably be higher priority than node names, since any attempt to use a preprocessor command as a node name would just get preprocessed instead. I can't say I'm super familiar with how tree-sitter works either though, unfortunately.

I also would not expect requiring whitespace as part of the token to be needed. `#define` on its own isn't technically valid, but it should probably parse as a token even if it isn't followed by anything I think?

Reply:
Yeah, I agree in principle, just not sure how to make it work is the problem.

Reply:
Based on the test case

```dts
/ {
    #include "macros.dtsi"
    #define FOO 0
};
```

The parser's debug output yields:

```
new_parse
process version:0, version_count:1, state:1, row:0, col:0
lex_internal state:85, row:0, column:0
  consume character:'/'
lexed_lookahead sym:identifier, size:1
shift state:192
process version:0, version_count:1, state:192, row:0, col:1
lex_internal state:0, row:0, column:1
  skip character:' '
  consume character:'{'
lexed_lookahead sym:{, size:2
shift state:5
process version:0, version_count:1, state:5, row:0, col:3
lex_internal state:30, row:0, column:3
  skip character:10
  skip character:' '
  skip character:' '
  skip character:' '
  skip character:' '
  consume character:'#'
  consume character:'i'
  consume character:'n'
  consume character:'c'
  consume character:'l'
  consume character:'u'
  consume character:'d'
  consume character:'e'
lexed_lookahead sym:_property_with_hash, size:13
shift state:184
process version:0, version_count:1, state:184, row:1, col:12
lex_internal state:85, row:1, column:12
  skip character:' '
lex_internal state:0, row:1, column:12
  skip character:' '
  consume character:'"'
lexed_lookahead sym:", size:2
detect_error
resume version:0
skip_token symbol:"

...
```

It normally says `lexed_lookahead sym:identifier, size:13` when parsing `#include` but I removed the alias for `property_identifier` to be certain that's what it was picking. If I update that rule:

```diff
-_property_with_hash: ($) => /[#0-9a-zA-Z,._+-]*#[#0-9a-zA-Z,._+-]*/,
+_property_with_hash: ($) => token(prec(-1, /[#0-9a-zA-Z,._+-]*#[#0-9a-zA-Z,._+-]*/)),
```

then the sample is parsed as expected, but of course this breaks the property _name with hash_ test. I don't understand precedence rules well enough to correctly apply it to the preproc rules in the context of a node.

Reply:
I guess is there a problem with proceeding with my PR as is?  It passes all tests and as you mentioned none of the preprocessor rules are valid without space after the directive.

Reply:
Might be getting closer to something... I can get tests to pass if the preproccessor rules are made with

```js
function preprocessor(command) {
	return token(prec(2, alias(new RegExp('#[ \t]*' + command), '#' + command)));
}
```

and no changes to the existing rules. However, this results in

```
Error in query file "highlights.scm"

Caused by:
    Query error at 4:6. Invalid node type
```

I haven't been able to use `tree-sitter highlight` to see how this should work or how it's impacted by the change.